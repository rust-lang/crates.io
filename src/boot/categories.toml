# This is where the categories available on crates.io are defined. To propose
# a change to the categories, send a pull request with your change made to this
# file.
#
# For help with TOML, see: https://github.com/toml-lang/toml
#
# Format:
#
# ```toml
# [slug]
# name = "Display name"
# description = "Give an idea of the crates that belong in this category."
#
#  [slug.categories.subcategory-slug]
#  name = "Subcategory display name, not including parent category display name"
#  description = "Give an idea of the crates that belong in this subcategory."
# ```
#
# Notes:
# - Slugs are the primary identifier. If you make a change to a category's slug,
#   crates that have been published with that slug will need to be updated to
#   use the new slug in order to stay in that category. If you only change
#   names and descriptions, those attributes CAN be updated without affecting
#   crates in that category.
# - Slugs are used in the path of URLs, so they should not contain spaces, `/`,
#   `@`, `:`, or `.`. They should be all lowercase.
#

[accessibility]
name = "Accessibility"
description = """
Assistive technology that helps overcome disabilities and impairments to make software usable by as many people as possible. \
"""

[aerospace]
name = "Aerospace"
description = """
Crates for aeronautics (within the atmosphere) and astronautics \
(in outer space) implications.\
"""

[aerospace.categories.drones]
name = "Drones"
description = """
Crates related to Multicopters, Fixed wing, VTOL (Vertical Takeoff and Landing) \
and Airships/Balloons.\
"""

[aerospace.categories.protocols]
name = "Aerospace protocols"
description = """
Crates of protocol implementations for aerospace implications.\
"""

[aerospace.categories.simulation]
name = "Aerospace simulations"
description = """
Crates related to any kind of simulations used in aerospace - fluids, \
aerodynamics, etc.\
"""

[aerospace.categories.space-protocols]
name = "Space protocols"
description = """
Protocol implementations for implications in space like CCSDS.\
"""

[aerospace.categories.unmanned-aerial-vehicles]
name = "Unmanned aerial vehicles"
description = """
Crates related to Unmanned aerial vehicles like Multicopters, Fixed wing, \
VTOL (Vertical Takeoff and Landing), Airships/Balloons, Rovers, Boats, \
Submersibles.\
"""

[algorithms]
name = "Algorithms"
description = """
Rust implementations of core algorithms such as hashing, sorting, \
searching, and more.\
"""

[api-bindings]
name = "API bindings"
description = """
Idiomatic wrappers of specific APIs for convenient access from \
Rust. Includes HTTP API wrappers as well. Non-idiomatic or unsafe \
bindings can be found in External FFI bindings.\
"""

[asynchronous]
name = "Asynchronous"
description = """
Crates to help you deal with events independently of the main program \
flow, using techniques like futures, promises, waiting, or eventing.\
"""

[authentication]
name = "Authentication"
description = """
Crates to help with the process of confirming identities.\
"""

[caching]
name = "Caching"
description = """
Crates to store the results of previous computations in order to reuse \
the results.\
"""

[compilers]
name = "Compilers"
description = """
Compiler implementations, including interpreters and transpilers.\
"""

[command-line-interface]
name = "Command-line interface"
description = """
Crates to help create command line interfaces, such as argument \
parsers, line-editing, or output coloring and formatting.\
"""

[command-line-utilities]
name = "Command line utilities"
description = """
Applications to run at the command line.\
"""

[compression]
name = "Compression"
description = """
Algorithms for making data smaller.\
"""

[computer-vision]
name = "Computer vision"
description = """
Crates for comprehending the world from video or images.\
"""

[config]
name = "Configuration"
description = """
Crates to facilitate configuration management for applications.\
"""

[concurrency]
name = "Concurrency"
description = """
Crates for implementing concurrent and parallel computation.\
"""

[cryptography]
name = "Cryptography"
description = """
Algorithms intended for securing data.\
"""

[cryptography.categories.cryptocurrencies]
name = "Cryptocurrencies"
description = """
Crates for digital currencies, wallets, and distributed ledgers.\
"""

[database]
name = "Database interfaces"
description = """
Crates to interface with database management systems.\
"""

[database-implementations]
name = "Database implementations"
description = """
Databases allow clients to store and query large amounts of data in an \
efficient manner. This category is for database management systems \
implemented in Rust.\
"""

[data-structures]
name = "Data structures"
description = """
Rust implementations of particular ways of organizing data suited for \
specific purposes.\
"""

[date-and-time]
name = "Date and time"
description = """
Crates to manage the inherent complexity of dealing with the fourth \
dimension.\
"""

[development-tools]
name = "Development tools"
description = """
Crates that provide developer-facing features such as testing, debugging, \
linting, performance profiling, autocompletion, formatting, and more.\
"""

[development-tools.categories.build-utils]
name = "Build Utils"
description = """
Utilities for build scripts and other build time steps.\
"""

[development-tools.categories.cargo-plugins]
name = "Cargo plugins"
description = """
Subcommands that extend the capabilities of Cargo.\
"""

[development-tools.categories.debugging]
name = "Debugging"
description = """
Crates to help you figure out what is going on with your code such as \
logging, tracing, or assertions.\
"""

[development-tools.categories.ffi]
name = "FFI"
description = """
Crates to help you better interface with other languages. This \
includes binding generators and helpful language constructs.\
"""

[development-tools.categories.procedural-macro-helpers]
name = "Procedural macro helpers"
description = """
Crates to help you write procedural macros in Rust.
"""

[development-tools.categories.profiling]
name = "Profiling"
description = """
Crates to help you figure out the performance of your code.\
"""

[development-tools.categories.testing]
name = "Testing"
description = """
Crates to help you verify the correctness of your code.\
"""

[email]
name = "Email"
description = """
Crates to help with sending, receiving, formatting, and parsing email.\
"""

[embedded]
name = "Embedded development"
description = """
Crates that are primarily useful on embedded devices or \
without an operating system.
"""

[emulators]
name = "Emulators"
description = """
Emulators allow one computer to behave like another, often to allow \
running software that is not natively available on the host \
computer. Video game systems are commonly emulated.\
"""

[encoding]
name = "Encoding"
description = """
Encoding and/or decoding data from one data format to another.\
"""

[external-ffi-bindings]
name = "External FFI bindings"
description = """
Direct Rust FFI bindings to libraries written in other languages; \
often denoted by a -sys suffix. Safe and idiomatic wrappers are in \
the API bindings category.
"""

[filesystem]
name = "Filesystem"
description = """
Crates for dealing with files and filesystems.\
"""

[finance]
name = "Finance"
description = """
Crates for dealing with money. Accounting, trading, investments, taxes, banking and payment processing using government-backed currencies.\
"""

[game-engines]
name = "Game engines"
description = """
For crates that try to provide a \"one-stop-shop\" for \
all of your game development needs.\
"""

[game-development]
name = "Game development"
description = """
For crates that focus on some individual part of accelerating \
the development of games.\
"""

[games]
name = "Games"
description = """
Applications for fun and entertainment. If Rust the video game were \
implemented in Rust the programming language, it would belong in this \
category. Libraries to help create video games are in the \
Game engines category.\
"""

[graphics]
name = "Graphics"
description = """
Crates for graphics libraries and applications, including raster and vector \
graphics primitives such as geometry, curves, and color.\
"""

[gui]
name = "GUI"
description = """
Crates to help you create a graphical user interface.\
"""

[hardware-support]
name = "Hardware support"
description = """
Crates to interface with specific CPU or other hardware features.\
"""

[internationalization]
name = "Internationalization (i18n)"
description = """
Crates to help develop software capable of adapting to various \
languages and regions.\
"""

[localization]
name = "Localization (L10n)"
description = """
Crates to help adapting internationalized software to specific \
languages and regions.\
"""

[mathematics]
name = "Mathematics"
description = """
Crates with a mathematical aspect.
"""

[memory-management]
name = "Memory management"
description = """
Crates to help with allocation, memory mapping, garbage collection, \
reference counting, or interfaces to foreign memory managers.\
"""

[multimedia]
name = "Multimedia"
description = """
Crates that provide audio, video, and image processing or rendering \
engines.\
"""

[multimedia.categories.audio]
name = "Audio"
description = """
Crates that record, output, or process audio.
"""

[multimedia.categories.video]
name = "Video"
description = """
Crates that record, output, or process video.
"""

[multimedia.categories.images]
name = "Images"
description = """
Crates that process or build images.
"""

[multimedia.categories.encoding]
name = "Encoding"
description = """
Crates that encode or decode binary data in multimedia formats.
"""

[network-programming]
name = "Network programming"
description = """
Crates dealing with higher-level network protocols such as FTP, HTTP, \
or SSH, or lower-level network protocols such as TCP or UDP.\
"""

[no-std]
name = "No standard library"
description = """
Crates that are able to function without the Rust standard library.
"""

[no-std.categories.no-alloc]
name = "No dynamic allocation"
description = """
Crates that are able to function without the Rust alloc crate.\
"""

[os]
name = "Operating systems"
description = """
Bindings to operating system-specific APIs.\
"""

[os.categories.android-apis]
name = "Android APIs"
description = """
Bindings to Android-specific APIs.\
"""

[os.categories.linux-apis]
name = "Linux APIs"
description = """
Bindings to Linux-specific APIs.\
"""

[os.categories.freebsd-apis]
name = "FreeBSD APIs"
description = """
Bindings to FreeBSD-specific APIs.\
"""

[os.categories.macos-apis]
name = "macOS APIs"
description = """
Bindings to macOS-specific APIs.\
"""

[os.categories.unix-apis]
name = "Unix APIs"
description = """
Bindings to Unix-specific APIs.\
"""

[os.categories.windows-apis]
name = "Windows APIs"
description = """
Bindings to Windows-specific APIs.\
"""

[parser-implementations]
name = "Parser implementations"
description = """
Parsers implemented for particular formats or languages.\
"""

[parsing]
name = "Parsing tools"
description = """
Crates to help create parsers of binary and text \
formats. Format-specific parsers belong in other, more specific \
categories.\
"""

[rendering]
name = "Rendering"
description = """
Real-time or offline rendering of 2D or 3D graphics, \
usually with the help of a graphics card.\
"""

[rendering.categories.engine]
name = "Rendering engine"
description = """
High-level solutions for rendering on the screen.\
"""

[rendering.categories.graphics-api]
name = "Graphics APIs"
description = """
Crates that provide direct access to the hardware's or the operating \
system's rendering capabilities.\
"""

[rendering.categories.data-formats]
name = "Data formats"
description = """
Loading and parsing of data formats related to 2D or 3D rendering, like \
3D models or animation sheets.\
"""

[rust-patterns]
name = "Rust patterns"
description = """
Shared solutions for particular situations specific to programming in \
Rust.\
"""

[science]
name = "Science"
description = """
Crates related to solving problems involving physics, chemistry, \
biology, machine learning, geoscience, and other scientific fields.\
"""

[science.categories.geo]
name = "Geospatial"
description = """
Processing of spatial information, maps, navigation data, and geographic information systems.
"""

[science.categories.robotics]
name = "Robotics"
description = """
Crates related to robotics.\
"""

[science.categories.neuroscience]
name = "Neuroscience"
description = """
Crates for research tools and processing of data related to the brain and nervous system.\
"""

[science.categories.bioinformatics]
name = "Bioinformatics"
description = """
Crates for processing large-scale biological data.\
"""

[science.categories.bioinformatics.categories.genomics]
name = "Genomics"
description = """
Crates for processing genetic data, including sequences, abdundance, variants and analysis.\
"""

[science.categories.bioinformatics.categories.proteomics]
name = "Proteomics"
description = """
Crates for processing protein data including sequences, abdundance and analysis.\
"""

[science.categories.bioinformatics.categories.sequence-analysis]
name = "Sequence analysis"
description = """
Crates for processing biological sequences, including alignment, assembly, and annotation.\
"""

[simulation]
name = "Simulation"
description = """
Crates used to model or construct models for some activity, e.g. to.\
simulate a networking protocol.\
"""

[template-engine]
name = "Template engine"
description = """
Crates designed to combine templates with data to produce result \
documents, usually with an emphasis on processing text.\
"""

[text-editors]
name = "Text editors"
description = """
Applications for editing text.\
"""

[text-processing]
name = "Text processing"
description = """
Crates to deal with the complexities of human language when expressed \
in textual form.\
"""

[value-formatting]
name = "Value formatting"
description = """
Crates to allow an application to format values for display to a user, \
potentially adapting the display to various languages and regions.\
"""

[virtualization]
name = "Virtualization"
description = """
For the creation and management of virtual environments and resources
of any form including containerization systems.\
"""

[visualization]
name = "Visualization"
description = """
Ways to view data, such as plotting or graphing.\
"""

[wasm]
name = "WebAssembly"
description = """
Crates for use when targeting WebAssembly, or for manipulating WebAssembly.\
"""

[web-programming]
name = "Web programming"
description = """
Crates to create applications for the web.\
"""

[web-programming.categories.http-client]
name = "HTTP client"
description = """
Crates to make HTTP network requests.\
"""

[web-programming.categories.http-server]
name = "HTTP server"
description = """
Crates to serve data over HTTP.\
"""

[web-programming.categories.websocket]
name = "WebSocket"
description = """
Crates to communicate over the WebSocket protocol.\
"""
