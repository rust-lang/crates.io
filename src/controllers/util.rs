use crate::middleware::app::RequestApp;
use crate::middleware::log_request::RequestLogExt;
use crate::util::errors::{forbidden, AppResult};
use crate::util::BytesRequest;
use http::request::Parts;
use http::{header, Extensions, HeaderMap, HeaderValue, Method, Request, Uri, Version};

/// The Origin header (<https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin>)
/// is sent with CORS requests and POST requests, and indicates where the request comes from.
/// We don't want to accept authenticated requests that originated from other sites, so this
/// function returns an error if the Origin header doesn't match what we expect "this site" to
/// be: <https://crates.io> in production, or <http://localhost:port/> in development.
pub fn verify_origin(parts: &Parts) -> AppResult<()> {
    let headers = parts.headers();
    let allowed_origins = &parts.app().config.allowed_origins;

    let bad_origin = headers
        .get_all(header::ORIGIN)
        .iter()
        .find(|value| !allowed_origins.contains(value));

    if let Some(bad_origin) = bad_origin {
        let error_message =
            format!("only same-origin requests can be authenticated. got {bad_origin:?}");

        parts.request_log().add("cause", error_message);

        return Err(forbidden("invalid origin header"));
    }
    Ok(())
}

pub trait RequestPartsExt {
    fn method(&self) -> &Method;
    fn uri(&self) -> &Uri;
    fn version(&self) -> Version;
    fn headers(&self) -> &HeaderMap<HeaderValue>;
    fn extensions(&self) -> &Extensions;
}

impl RequestPartsExt for Parts {
    fn method(&self) -> &Method {
        &self.method
    }
    fn uri(&self) -> &Uri {
        &self.uri
    }
    fn version(&self) -> Version {
        self.version
    }
    fn headers(&self) -> &HeaderMap<HeaderValue> {
        &self.headers
    }
    fn extensions(&self) -> &Extensions {
        &self.extensions
    }
}

impl<B> RequestPartsExt for Request<B> {
    fn method(&self) -> &Method {
        self.method()
    }
    fn uri(&self) -> &Uri {
        self.uri()
    }
    fn version(&self) -> Version {
        self.version()
    }
    fn headers(&self) -> &HeaderMap<HeaderValue> {
        self.headers()
    }
    fn extensions(&self) -> &Extensions {
        self.extensions()
    }
}

impl RequestPartsExt for BytesRequest {
    fn method(&self) -> &Method {
        self.0.method()
    }
    fn uri(&self) -> &Uri {
        self.0.uri()
    }
    fn version(&self) -> Version {
        self.0.version()
    }
    fn headers(&self) -> &HeaderMap<HeaderValue> {
        self.0.headers()
    }
    fn extensions(&self) -> &Extensions {
        self.0.extensions()
    }
}
