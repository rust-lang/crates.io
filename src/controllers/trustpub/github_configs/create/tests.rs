use crate::tests::builders::CrateBuilder;
use crate::tests::util::{RequestHelper, Response, TestApp};
use anyhow::anyhow;
use bytes::Bytes;
use crates_io_database::models::token::{CrateScope, EndpointScope};
use crates_io_database::schema::{emails, trustpub_configs_github};
use crates_io_github::{GitHubError, GitHubUser, MockGitHubClient};
use diesel::prelude::*;
use diesel_async::RunQueryDsl;
use insta::{assert_json_snapshot, assert_snapshot};
use serde_json::json;

const URL: &str = "/api/v1/trusted_publishing/github_configs";

const CRATE_NAME: &str = "foo";

fn simple_github_mock() -> MockGitHubClient {
    let mut github_mock = MockGitHubClient::new();
    github_mock.expect_get_user().returning(|login, _| {
        Ok(GitHubUser {
            avatar_url: None,
            email: None,
            id: 42,
            login: login.into(),
            name: None,
        })
    });
    github_mock
}

async fn run_test(payload: impl Into<Bytes>) -> (TestApp, Response<()>) {
    async fn inner(payload: Bytes) -> (TestApp, Response<()>) {
        let (app, _client, cookie_client) = TestApp::full()
            .with_github(simple_github_mock())
            .with_user()
            .await;

        let mut conn = app.db_conn().await;

        CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
            .build(&mut conn)
            .await
            .unwrap();

        (app, cookie_client.post::<()>(URL, payload).await)
    }

    inner(payload.into()).await
}

#[tokio::test(flavor = "multi_thread")]
async fn test_happy_path() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let (app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    assert_snapshot!(app.emails_snapshot().await);

    let mut conn = app.db_conn().await;
    let config_ids = trustpub_configs_github::table
        .select(trustpub_configs_github::id)
        .get_results::<i32>(&mut conn)
        .await?;

    assert_eq!(config_ids.len(), 1);
    assert_eq!(config_ids[0], 1);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_happy_path_with_environment() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": "production",
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_empty_body() -> anyhow::Result<()> {
    let (_app, response) = run_test("").await;
    assert_snapshot!(response.status(), @"415 Unsupported Media Type");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Expected request with `Content-Type: application/json`"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_empty_json_object() -> anyhow::Result<()> {
    let (_app, response) = run_test("{}").await;
    assert_snapshot!(response.status(), @"422 Unprocessable Entity");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Failed to deserialize the JSON body into the target type: missing field `github_config` at line 1 column 2"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_owner() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "ยง$%&",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Invalid GitHub repository owner name"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_repo() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "@foo",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Invalid GitHub repository name"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_workflow() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "workflows/ci.json",
            "environment": null,
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Workflow filename must end with `.yml` or `.yaml`"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_environment() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": "",
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Environment name may not be empty (use `null` to omit)"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_unauthenticated() -> anyhow::Result<()> {
    let (app, client, cookie_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_user()
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"403 Forbidden");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"this action requires authentication"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_legacy_token_auth() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_token()
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth_with_trusted_publishing_scope() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_scoped_token(
            Some(vec![CrateScope::try_from(CRATE_NAME).unwrap()]),
            Some(vec![EndpointScope::TrustedPublishing]),
        )
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth_without_trusted_publishing_scope() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_scoped_token(
            Some(vec![CrateScope::try_from(CRATE_NAME).unwrap()]),
            Some(vec![EndpointScope::PublishUpdate]),
        )
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"403 Forbidden");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"this token does not have the required permissions to perform this action"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth_with_wrong_crate_scope() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_scoped_token(
            Some(vec![CrateScope::try_from("other-crate").unwrap()]),
            Some(vec![EndpointScope::TrustedPublishing]),
        )
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"403 Forbidden");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"this token does not have the required permissions to perform this action"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth_with_wildcard_crate_scope() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_scoped_token(
            Some(vec![CrateScope::try_from("*").unwrap()]),
            Some(vec![EndpointScope::TrustedPublishing]),
        )
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_missing_crate() -> anyhow::Result<()> {
    let (_app, _client, cookie_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_user()
        .await;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = cookie_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"404 Not Found");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"crate `foo` does not exist"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_non_owner() -> anyhow::Result<()> {
    let (app, _client, cookie_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_user()
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let other_client = app.db_new_user("other_user").await;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = other_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"You are not an owner of this crate"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_unknown_github_user() -> anyhow::Result<()> {
    let mut github_mock = MockGitHubClient::new();
    github_mock
        .expect_get_user()
        .returning(|login, _| Err(GitHubError::NotFound(anyhow!("User {} not found", login))));

    let (app, _client, cookie_client) = TestApp::full().with_github(github_mock).with_user().await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = cookie_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Unknown GitHub user or organization"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_github_error() -> anyhow::Result<()> {
    let mut github_mock = MockGitHubClient::new();
    github_mock
        .expect_get_user()
        .returning(|_, _| Err(GitHubError::Other(anyhow!("Internal Server Error"))));

    let (app, _client, cookie_client) = TestApp::full().with_github(github_mock).with_user().await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = cookie_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"500 Internal Server Error");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Internal Server Error"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_unverified_email() -> anyhow::Result<()> {
    let (app, _client, cookie_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_user()
        .await;

    let mut conn = app.db_conn().await;

    diesel::update(emails::table.filter(emails::user_id.eq(cookie_client.as_model().id)))
        .set(emails::verified.eq(false))
        .execute(&mut conn)
        .await?;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = cookie_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"403 Forbidden");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"You must verify your email address to create a Trusted Publishing config"}]}"#);

    Ok(())
}
