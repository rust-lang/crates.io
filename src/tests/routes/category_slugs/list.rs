use crate::tests::new_category;
use crate::tests::util::{RequestHelper, TestApp};
use crates_io_database::schema::categories;
use diesel::{insert_into, RunQueryDsl};
use insta::assert_json_snapshot;
use serde_json::Value;

#[tokio::test(flavor = "multi_thread")]
async fn category_slugs_returns_all_slugs_in_alphabetical_order() {
    let (app, anon) = TestApp::init().empty();
    let mut conn = app.db_conn();

    let cats = vec![
        new_category("Foo", "foo", "For crates that foo"),
        new_category("Bar", "bar", "For crates that bar"),
    ];

    insert_into(categories::table)
        .values(cats)
        .execute(&mut conn)
        .unwrap();

    let response: Value = anon.get("/api/v1/category_slugs").await.good();
    assert_json_snapshot!(response);
}
