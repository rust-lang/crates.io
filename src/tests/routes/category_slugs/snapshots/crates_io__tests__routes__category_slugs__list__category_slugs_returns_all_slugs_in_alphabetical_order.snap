---
source: src/tests/routes/category_slugs/list.rs
expression: response
---
{
  "category_slugs": [
    {
      "description": "For crates that bar",
      "id": "bar",
      "slug": "bar"
    },
    {
      "description": "For crates that foo",
      "id": "foo",
      "slug": "foo"
    }
  ]
}
